src/pytorch/CFConv.cpp

```diff
@@ -0,0 +1,300 @@
+#include "CpuCFConv.h"
+#include "CFConvNeighbors.h"
+
+#ifdef ENABLE_CUDA
+#include <stdexcept>
+#include <cuda_runtime.h>
+// #include <c10/cuda/CUDAStream.h>
+#include "CudaCFConv.h"
+
+#define CHECK_CUDA_RESULT(result) \
+    if (result != cudaSuccess) { \
+        throw std::runtime_error(std::string("Encountered error ")+cudaGetErrorName(result)+" at "+__FILE__+":"+std::to_string(__LINE__));\
+    }
+#endif
+
+namespace NNPOps {
+namespace CFConv {
+
+class Holder;
+using Activation = ::CFConv::ActivationFunction;
+using Context = torch::autograd::AutogradContext;
+using HolderPtr = torch::intrusive_ptr<Holder>;
+using Neighbors = NNPOps::CFConvNeighbors::Holder;
+using NeighborsPtr = torch::intrusive_ptr<Neighbors>;
+using std::string;
+using torch::autograd::tensor_list;
+using torch::Device;
+using torch::IValue;
+using torch::optional;
+using torch::Tensor;
+using torch::TensorOptions;
+
+class Holder : public torch::CustomClassHolder {
+public:
+    Holder(double gaussianWidth,
+           const string& activation,
+           const Tensor& weights1,
+           const Tensor& biases1,
+           const Tensor& weights2,
+           const Tensor& biases2) :
+
+        torch::CustomClassHolder(),
+        gaussianWidth(gaussianWidth),
+        activation(activation),
+        // Note: weights and biases have to be in the CPU memory
+        weights1(weights1.to(torch::kFloat32).cpu().clone()),
+        biases1(biases1.to(torch::kFloat32).cpu().clone()),
+        weights2(weights2.to(torch::kFloat32).cpu().clone()),
+        biases2(biases2.to(torch::kFloat32).cpu().clone()),
+        device(torch::kCPU)
+    {};
+
+    Tensor forward(const IValue& neighbors_, const Tensor& positions, const Tensor& input) {
+
+        neighbors = neighbors_.toCustomClass<Neighbors>(); // save for the backward pass
+
+        this->positions = positions; // save for the backward pass
+        if (positions.scalar_type() != torch::kFloat32)
+            throw std::runtime_error("The type of \"positions\" has to be float32");
+        if (positions.dim() != 2)
+            throw std::runtime_error("The shape of \"positions\" has to have 2 dimensions");
+        if (positions.size(1) != 3)
+            throw std::runtime_error("The size of the 2nd dimension of \"positions\" has to be 3");
+
+        this->input = input; // save for the backward pass
+        if (input.device() != positions.device())
+            throw std::runtime_error("The device of \"input\" and \"positions\" has to be the same");
+        if (input.scalar_type() != torch::kFloat32)
+            throw std::runtime_error("The type of \"input\" has to be float32");
+        if (input.dim() != 2)
+            throw std::runtime_error("The shape of \"input\" has to have 2 dimensions");
+        if (input.size(0) != positions.size(0))
+            throw std::runtime_error("The size of the 1nd dimension of \"input\" has to be equal to the 1st dimension of \"positions\"");
+
+        if(!impl) {
```

Comment:
I'm not certain if this is guaranteed to work correctly or if you could get uninitialized memory.  I would add `impl(nullptr)` to the constructor initialization list to be sure.  I've been pondering the `shared_ptr` documentation, and it's ambiguous on this.

Reply:
`std::shared_prt()` and `std::shared_ptr(nullptr)` are equivalent (https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr).